### PR TITLE
fix(report): restore patient appointment list export

### LINKED ISSUE #3346: Fix Patient List by Appointment Time export returning 404
## Problem

The **Export** button under **Administration → Reports → Patient List by Appointment Time** does not download `patientlist.txt`.

Submitting valid filters sends a request such as:

```text
GET /carlos/patientlistbyappt?provider_no=all&date_from=2026-08-07&date_to=2026-08-10&CSRF-TOKEN=...
```

The server responds with HTTP 404 and the CARLOS error page. No attachment is downloaded.

Runtime logs report:

```text
There is no Action mapped for namespace [/] and action name [patientlistbyappt]
```

`web.xml` maps `/patientlistbyappt` to `io.github.carlos_emr.carlos.report.data.PatientListByAppt`, and `HttpMethodGuardFilter.NON_STRUTS_PREFIXES` lists the route, but the request is still handled by the Struts dispatcher instead of the servlet.

## Additional defect

Once routing is restored, valid appointments with a null `type` can cause an exception here:

```java
escapeCsv(a.getType().replaceAll("\r\n", ""))
```

The exporter must handle null appointment fields safely.

## Steps to reproduce

1. Open **Administration → Reports → Patient List by Appointment Time**.
2. Select **All Doctors**.
3. Enter a date range containing appointments.
4. Click **Export**.
5. Observe the HTTP 404 error page and absence of a download.

## Expected behavior

The request reaches the Patient List export handler and downloads a UTF-8 `patientlist.txt` file containing the matching appointments.

## Acceptance criteria

- `/patientlistbyappt` is routed to the intended authenticated export handler rather than the Struts action dispatcher.
- Clicking **Export** returns HTTP 200 with an attachment named `patientlist.txt`.
- All Doctors exports appointments across providers within the inclusive date range.
- Selecting an individual doctor exports only that provider's appointments.
- Exported rows include patient name, phone numbers, appointment time/date/type, provider, and location in the documented order.
- CSV escaping and spreadsheet-formula protection remain intact.
- Null appointment type, location, phone, and other optional values export as empty fields rather than causing HTTP 500.
- Empty result sets produce a valid empty export rather than an error page.
- Unauthenticated requests remain rejected according to the download endpoint policy.
- Add Playwright coverage that submits the visible form, captures the download, and validates filename and representative content.
- Add servlet/request tests covering route resolution, provider/date filters, empty results, and null optional fields.

### PR DESCRIPTION
**Pull request by Ben-Heerema**
## Description

- excludes the extensionless `/patientlistbyappt` servlet route from the Struts dispatcher so the existing authenticated `web.xml` mapping receives export requests
- enforces `_report` or `_admin.reporting` read privilege at the direct servlet boundary before querying PHI
- strictly requires provider and `YYYY-MM-DD` date filters, rejects reversed ranges, and prevents invalid input from widening into an unbounded export
- submits the visible form with POST so `CSRF-TOKEN` remains out of URLs, logs, and browser history
- restores dependency-free browser validation for required, malformed, impossible, and reversed dates
- HTML-encodes provider IDs and names in the visible selector to prevent stored markup injection
- exports null appointment types as empty fields, removes embedded CR/LF characters safely, and protects CSV consumers from formula injection
- streams a narrow scalar projection through a forward-only database cursor so large date ranges do not materialize full entities or the complete result set in memory
- writes synchronous, PHI-free audit records for successful and failed export attempts with provider filter, date range, row count, outcome, and exception type
- restricts the Playwright harness's default credentials and relaxed TLS handling to loopback targets; intentional non-loopback runs require HTTPS, explicit credentials, and opt-in
- adds focused servlet/config regression coverage, a DAO cursor integration test, and a Playwright check that validates client-side dates, POST downloads, provider filtering, empty results, null fields, browser errors, and anonymous rejection

## Related Issues

Fixes #3346

## How Was This Tested?

- `mvn -Dtest='PatientListByApptExportUnitTest,OscarAppointmentDaoQueryIntegrationTest$FindPatientAppointments#shouldStreamProjection_whenAppointmentsExist' test` — 18 tests passed (17 servlet/config unit tests and 1 Hibernate cursor integration test)
- `mvn pmd:pmd -DskipTests` — completed successfully with no findings in changed code
- focused SpotBugs + FindSecBugs analysis of `PatientListByAppt`, `OscarAppointmentDaoImpl`, and `PatientAppointmentExportRow` — no findings in changed code (four pre-existing findings remain in unrelated DAO methods)
- `mvn -DskipTests package war:exploded` — build and Checkstyle passed
- `mvn -Pjspc jspc:compile@jspc` — all 973 JSPs compiled with 0 errors
- `node --check scripts/patient-list-by-appointment-export-playwright-checks.js`
- harness boundary checks confirmed RFC1918/non-loopback targets are rejected without explicit opt-in, HTTPS, and credentials
- live Playwright run against local CARLOS/Tomcat with MySQL — all-doctors (3 rows), provider 9 (1), provider 999998 (2), empty range (0), null appointment type, POST/download envelope, unauthenticated 401, blank/malformed/impossible/reversed client dates, and browser/server error checks passed with no findings
- live database verification confirmed four matching `export` audit records with correct filters and row counts and no patient data
- `git diff --check`

## Checklist

- [x] My commits are signed off for the [DCO](https://developercertificate.org/) (`git commit -s`)
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/) format
- [x] I have not included any patient data (PHI) in this PR
- [x] I have added tests for new functionality
- [x] I have read the [contributing guide](CONTRIBUTING.md)

### CODE REVIEW

**Comment by Ben-Heerema** [comment]
@coderabbitai review

**Comment by Ben-Heerema** [comment]
@sourcery-ai review

File: scripts/patient-list-by-appointment-export-playwright-checks.js [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+  // the value directly and dispatching input/change gives jQuery Validate the
+  // same signal without ever opening the calendar.
+  await page.evaluate(({ from, to }) => {
+    for (const [id, value] of [['date_from', from], ['date_to', to]]) {
+      const input = document.getElementById(id);
+      input.value = value;
+      input.dispatchEvent(new Event('input', { bubbles: true }));
+      input.dispatchEvent(new Event('change', { bubbles: true }));
+    }
+  }, { from: dateFrom, to: dateTo });
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/patient-list-by-appointment-export-playwright-checks.js]: Addressed in 87cc7e85ce: removed the argument-bearing `page.evaluate` call entirely. Date values are now set with Playwright locator APIs, validated with `inputValue()`/`checkValidity()`, and the flatpickr widget is closed using a fixed no-argument callback. The full live export workflow passes with the expected provider filters and row counts. Waiting for the refreshed Semgrep PRO scan before resolving this thread.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

File: src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java [hunk 2]
```diff
@@ -119,13 +88,224 @@ protected void processRequest(HttpServletRequest request, HttpServletResponse re
...
+
+        int[] rowCount = {0};
+        String outcome = "error";
+        String errorType = null;
+
+        try {
+            OscarAppointmentDao dao = getAppointmentDao();
+            PrintWriter pw = new PrintWriter(new OutputStreamWriter(
+                    response.getOutputStream(), StandardCharsets.UTF_8), true);
+            dao.streamPatientAppointments(providerNo, from, to, row -> {
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java]: Fixed in 2c12c8971b. The DAO now streams into an owner-only temporary spool file; response headers and bytes are emitted only after the DAO cursor and spool writer complete. A regression test verifies that a DAO failure after one row returns HTTP 500 with no partial attachment.

File: src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java [hunk 3]
```diff
@@ -119,13 +88,224 @@ protected void processRequest(HttpServletRequest request, HttpServletResponse re
...
+        }
+        return firstName + " " + lastName;
+    }
+
+    private void rejectExport(HttpServletResponse response, LoggedInInfo loggedInInfo,
+                              int status, String message, String reason) throws IOException {
+        response.sendError(status, message);
+        // Rejected parameters are deliberately omitted: malformed attacker-controlled
+        // values must not become PHI or log-injection content in the audit record.
+        auditExport(loggedInInfo, null, null, null, 0, "rejected", reason);
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java]: Fixed in 2c12c8971b. Rejection audits now retain only provider/date fields that have passed validation, with dates canonicalized before persistence; malformed raw values remain excluded.

File: src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java [hunk 4]
```diff
@@ -119,13 +88,224 @@ protected void processRequest(HttpServletRequest request, HttpServletResponse re
...
+     * can commit an empty success response before {@link HttpServletResponse#sendError}
+     * can replace it.
+     */
+    @SuppressWarnings("java:S2093")
+    private void processExportRequest(HttpServletRequest request, HttpServletResponse response)
+            throws IOException {
+        LoggedInInfo loggedInInfo = LoggedInInfo.getLoggedInInfoFromSession(request);
+        if (loggedInInfo == null) {
+            UnauthenticatedRejectionResolver.rejectUnauthenticatedRequest(request, response);
+            auditUnauthenticatedExport(request.getRemoteAddr());
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java]: Resolved in 2c12c8971b. The anonymous-path synchronous database audit was removed to prevent unauthenticated write amplification. The centralized UnauthenticatedRejectionResolver remains responsible for emitting the sanitized application-level rejection event.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

File: src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java [hunk 5]
```diff
@@ -119,13 +92,254 @@ protected void processRequest(HttpServletRequest request, HttpServletResponse re
...
+            response.setContentType("text/plain; charset=UTF-8");
+            response.setHeader("Content-disposition", "attachment; filename=patientlist.txt");
+            Files.copy(spoolFile, response.getOutputStream());
+            outcome = "success";
+        } catch (IOException | RuntimeException e) {
+            errorType = e.getClass().getSimpleName();
+            throw e;
+        } finally {
+            deleteExportSpoolFile(spoolFile);
+            auditExport(loggedInInfo, validatedScope.providerFilter(),
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/report/data/PatientListByAppt.java]: Fixed in 4755f748ff. Audit persistence now runs through a guarded helper for export-finally, authorization-failure, and rejection paths; an audit-store exception is logged and cannot replace the primary failure. The response-write regression now injects an audit outage and verifies the original IOException still propagates.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@yingbull ready for review

**Review by yingbull — APPROVED** [review]
